autogen/coding/embedded_ipython_code_executor.py

```diff
@@ -0,0 +1,221 @@
+        if self.kernel_name not in KernelSpecManager().find_kernel_specs():
+            raise ValueError(
+                f"Kernel {self.kernel_name} is not installed. "
+                "Please first install it with "
+                f"`python -m ipykernel install --user --name {self.kernel_name}`."
+            )
+        self._kernel_manager = KernelManager(kernel_name=self.kernel_name)
+        self._kernel_manager.start_kernel()
+        self._kernel_client = self._kernel_manager.client()
+        self._kernel_client.start_channels()
+        self._timeout = self.timeout
+
+    @property
+    def user_capability(self) -> "EmbeddedIPythonCodeExecutor.UserCapability":
+        """Export a user capability that can be added to an agent."""
+        return EmbeddedIPythonCodeExecutor.UserCapability()
+
+    @property
+    def code_extractor(self) -> CodeExtractor:
+        """Export a code extractor that can be used by an agent."""
+        return MarkdownCodeExtractor()
+
+    def execute_code_blocks(self, code_blocks: List[CodeBlock]) -> IPythonCodeResult:
+        """Execute a list of code blocks and return the result.
+
+        This method executes a list of code blocks as cells in an IPython kernel
+        managed by this class.
+        See: https://jupyter-client.readthedocs.io/en/stable/messaging.html
+        for the message protocol.
+
+        Args:
+            code_blocks (List[CodeBlock]): A list of code blocks to execute.
+
+        Returns:
+            IPythonCodeResult: The result of the code execution.
+        """
+        self._kernel_client.wait_for_ready()
+        outputs = []
+        output_files = []
+        for code_block in code_blocks:
+            code = self._process_code(code_block.code)
+            self._kernel_client.execute(code, store_history=True)
+            while True:
+                try:
+                    msg = self._kernel_client.get_iopub_msg(timeout=self._timeout)
+                    msg_type = msg["msg_type"]
+                    content = msg["content"]
+                    if msg_type in ["execute_result", "display_data"]:
+                        for data_type, data in content["data"].items():
+                            if data_type == "text/plain":
+                                # Output is a text.
+                                outputs.append(data)
+                            elif data_type.startswith("image/"):
+                                # Output is an image.
+                                path = self._save_image(data)
+                                outputs.append(f"Image data saved to {path}")
+                                output_files.append(path)
+                            elif data_type == "text/html":
+                                # Output is an html.
+                                path = self._save_html(data)
+                                outputs.append(f"HTML data saved to {path}")
+                                output_files.append(path)
```

Comment:
Correct. I hoping do add those clean up routines as future work. Right now, I have seen people using existing code execution utils complaining about where to find generated code files, as often the files are removed automatically if they were not named by the user or the LLM. So in this PR for command line executor I explicitly save all the files -- disk space is typically not an issue.

I am hoping as people use it, we will discover what is the best way to do the cleanup. Currently, I would say the easiest way is to have each code executor using it's own working directory to avoid cluttering.

Reply:
@AaronWard would you like to resolve this? We can improve on the clean up in the future.